src/background/navigation-instrument.ts

```diff
@@ -1,70 +1,156 @@
+      navigation.before_navigate_event_ordinal = incrementedEventOrdinal();
+      navigation.before_navigate_time_stamp = new Date(
+        details.timeStamp,
+      ).toISOString();
+      pendingNavigation.resolveOnBeforeNavigateEventNavigation(navigation);
+    };
+    browser.webNavigation.onBeforeNavigate.addListener(
+      this.onBeforeNavigateListener,
+    );
     this.onCommittedListener = async (
       details: WebNavigationOnCommittedEventDetails,
     ) => {
-      const tab =
-        details.tabId > -1
-          ? await browser.tabs.get(details.tabId)
-          : {
-              windowId: undefined,
-              incognito: undefined,
-              cookieStoreId: undefined,
-              openerTabId: undefined,
-              width: undefined,
-              height: undefined,
-            };
-      const window = tab.windowId
-        ? await browser.windows.get(tab.windowId)
-        : { width: undefined, height: undefined, type: undefined };
-      const navigation: Navigation = {
-        crawl_id: crawlID,
-        incognito: boolToInt(tab.incognito),
-        extension_session_uuid: extensionSessionUuid,
-        process_id: details.processId,
-        window_id: tab.windowId,
-        tab_id: details.tabId,
-        tab_opener_tab_id: tab.openerTabId,
-        frame_id: details.frameId,
-        parent_frame_id: (details as any).parent_frame_id, // An undocumented property
-        window_width: window.width,
-        window_height: window.height,
-        window_type: window.type,
-        tab_width: tab.width,
-        tab_height: tab.height,
-        tab_cookie_store_id: escapeString(tab.cookieStoreId),
-        uuid: makeUUID(),
-        url: escapeString(details.url),
-        transition_qualifiers: escapeString(
-          JSON.stringify(details.transitionQualifiers),
-        ),
-        transition_type: escapeString(details.transitionType),
-        committed_time_stamp: new Date(details.timeStamp).toISOString(),
-      };
+      const navigationId = NavigationInstrument.navigationId(
+        details.processId,
+        details.tabId,
+        details.frameId,
+      );
+      const navigation: Navigation = await transformWebNavigationBaseEventDetailsToOpenWPMSchema(
+        crawlID,
+        details,
+      );
+      navigation.transition_qualifiers = escapeString(
+        JSON.stringify(details.transitionQualifiers),
+      );
+      navigation.transition_type = escapeString(details.transitionType);
       navigation.committed_event_ordinal = incrementedEventOrdinal();
+      navigation.committed_time_stamp = new Date(
+        details.timeStamp,
+      ).toISOString();
+
+      // include attributes from the corresponding onBeforeNavigation event
+      const pendingNavigation = this.getPendingNavigation(navigationId);
+      if (pendingNavigation) {
+        pendingNavigation.resolveOnCommittedEventNavigation(navigation);
+        const resolved = await pendingNavigation.resolvedWithinTimeout(1000);
```

Comment:
Under what conditions will this not resolve?

Reply:
If the onBeforeNavigate event never was recorded in the first place. "Should never happen" - but then again I don't want the instrumentation get stuck forever (until the next browser restart) in case it happens.